### PR TITLE
Recipe-app-development-pr

### PR DESCRIPTION
**Pull request by mmsesay**
In the project, @redwing555 and @mmsesay worked on implementing the following bullet points below in the development of this recipe app.

- A login page.
- A registration page.
- A food list (with all CRUD implementation, except for 'update').
- A recipes list (with all CRUD implementation, except for 'update').
- A general shopping list view (all missing food for all your recipes and total price).
- Solve N+1 query problem
- Wrote unit tests and integration tests

### CODE REVIEW

File: app/views/devise/registrations/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    <%= f.label :password %>
+    <% if @minimum_password_length %>
+    <em>(<%= @minimum_password_length %> characters minimum)</em>
+    <% end %><br />
+    <%= f.password_field :password, autocomplete: "new-password", class: "form-control my-3" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password", class: "form-control my-3" %>
```
Comment (Dipeshtwis) [inline app/views/devise/registrations/new.html.erb:29]: - [ ] From the UI, I can't able to signup, It's giving me this error which i have attached below. Please fix this. I have checked your application by making the user through the command line.

Screenshot from 2022-02-04 00-14-59

File: app/controllers/food_controller.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    @foods = Food.where.missing(:recipe_foods)
+  end
+
+  def food_params
+    params.require(:food).permit(:name, :measurement_unit, :price)
+  end
+
+  def recipe_param
+    params.require(:food).permit(:recipe_id)
+  end
```
Comment (Dipeshtwis) [inline app/controllers/food_controller.rb:43]: - [ ] You have used these methods in a different method in the same class so it's better to make these two methods `private` to make your app more secure.

**Review by Dipeshtwis — CHANGES_REQUESTED** [review]
Hi @mmsesay and @redwing555, 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


 ## Required Changes ♻️

**_Check the comments under the review_**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

- N.A.

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**_Dipesh Kumar_**


------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/food_controller.rb (hunk 2 above)
Comment (oloomoses) [reply app/controllers/food_controller.rb:45]: Well done on making this suggested change

File: app/views/devise/registrations/new.html.erb (hunk 1 above)
Comment (oloomoses) [reply app/views/devise/registrations/new.html.erb:29]: - [ ] The error has been fixed. However, I was still not able to log in after registering as a user. I keep getting the below message on the login screen
Screenshot from 2022-02-04 01-33-49
Kindly fix it and ensure that a user is redirected after a successful login.

**Review by oloomoses — CHANGES_REQUESTED** [review]
Hi team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
- [ ] Some tests are failing. this might be a result of changes made from the previous review. It would be better to look into this and fix the failing tests.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

_Slack: Oloo Moses_
------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb (hunk 1 above)
Comment (redwing555) [reply app/views/devise/registrations/new.html.erb:29]: Hello @oloomoses, we are using email confirmation that's why you weren't able to log In, however, I changed the root route now and the send notification for email confirmation can be seen.

**Review by Emmyn5600 — APPROVED** [review]
STATUS: APPROVED  :partying_face: 

Wooow👏, Your project is complete!  There is nothing else to say other than... it's time to merge it. 👏

Congratulations! 🎉 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**